transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```
@@ -916,6 +916,15 @@ private static void notifyOutboundHandlerException(Throwable cause, ChannelPromi
         PromiseNotificationUtil.tryFailure(promise, cause, logger);
     }
 
+    private void handleFatalOutboundHandlerException(Throwable cause) {
```

@trustin @vietj @chrisvest @yawkat actually I wonder if we should always do this for any outbound operation. The user is supposed to handle exceptions in their code by notify the promise. At the moment we try to fail the promise for them when something throws but imho the user should just handle it and fix their code. WDYT ?

---

I mean, even if you remove `throws Exception`, you still need to handle Error and RuntimeException. For those types, likely the best way will still be to fail the promise. So I don't really see a point in removing the `throws Exception` either.

---

By removing it at least people will be aware that they will need to handle exceptions and release messages etc . Having a promise as argument while having throws Exception is just confusing ... we had leaks before because people did not handle Exceptions. 

That's also the reason why I think we should log a warning and close the channel if people not correctly use the promise to propagate stuff back in general. Or maybe just log and try to fail the promise (which the user also might already have completed).

Generally speaking I would argue that have any type of exception be thrown is a user error if not catched by the user and fail the promise

---

Also just as a side-note... We removed the whole `throws Exception` in the old "main" as well. Here the question was more about what we should do if some of these methods throws an `Error` or `RuntimeException`.

---

Unclosed resources are a good reason, but that logic really also applies to read method, not just write. But I guess there we don't have a promise readily available.

---

yeah we don't have a promise in `flush(...)` and also not in `read(...)`. For both of these methods I really don't see why any of them should throw so it's a bit easier to decide to log and close. For others that take a `ChannelPromise` I am not sure yet.

That said I still kind of think that if a method takes a `ChannelPromise` it should never let an exception be thrown in the first place and if it does it's a user error. But maybe that is a question for a followup issue.

---

I think if there's a channel promise, we should notify through that, and only use this fallback for `read` and `flush`.

---

Yep we should but i still think we may want to consider logging a warn ... but this can be a follow discussion